Exec tests are flaky

These tests often fail on CI.

Eg. in https://github.com/moby/buildkit/pull/2082/checks?check_run_id=2666619721


```
--- FAIL: TestContainerdWorkerIntegration (0.05s)
    --- FAIL: TestContainerdWorkerIntegration/TestContainerdWorkerExec/worker=containerd-snapshotter-stargz (1.79s)
        common.go:78: Stdout: hello
        common.go:79: Stderr: 
        common.go:114: Stdout: 
        common.go:115: Stderr: 
        common.go:118: 
            	Error Trace:	common.go:118
            	Error:      	"" does not contain "1 sleep"
            	Test:       	TestContainerdWorkerIntegration/TestContainerdWorkerExec/worker=containerd-snapshotter-stargz
        run.go:166: 
            	Error Trace:	run.go:166
            	            				panic.go:563
            	            				testing.go:653
            	            				common.go:118
            	Error:      	Received unexpected error:
            	            	unlinkat /tmp/bktest_containerd399201633/state/io.containerd.runtime.v2.task/buildkit-test/wd4kni1y3hihiyz7gawutpz9l/rootfs: device or resource busy
            	Test:       	TestContainerdWorkerIntegration/TestContainerdWorkerExec/worker=containerd-snapshotter-stargz
        sandbox.go:221: stdout: /usr/bin/containerd-stargz-grpc
        sandbox.go:221: stderr: /usr/bin/containerd-stargz-grpc
        sandbox.go:224: > startCmd 2021-05-25 15:07:46.976737787 +0000 UTC m=+0.085574605 [containerd-stargz-grpc --log-level debug --address /tmp/bktest_containerd_stargz_grpc426742092/containerd-stargz-grpc.sock --root /tmp/bktest_containerd_stargz_grpc426742092/root]
        sandbox.go:224: > sending sigterm 2021-05-25 15:07:48.736075854 +0000 UTC m=+1.844912672
        sandbox.go:224: {"level":"info","msg":"Got terminated","time":"2021-05-25T15:07:48.736144959Z"}
        sandbox.go:224: {"level":"info","msg":"Exiting","time":"2021-05-25T15:07:48.736202963Z"}
        sandbox.go:224: > stopped 2021-05-25 15:07:48.737592452 +0000 UTC m=+1.846429270 exit status 0 0
        sandbox.go:221: stdout: /usr/bin/containerd
        sandbox.go:221: stderr: /usr/bin/containerd
        sandbox.go:224: > startCmd 2021-05-25 15:07:47.042175997 +0000 UTC m=+0.151012715 [containerd --config /tmp/bktest_containerd399201633/config.toml]
        sandbox.go:224: time="2021-05-25T15:07:47Z" level=warning msg="deprecated version : `1`, please switch to version `2`"
        sandbox.go:224: time="2021-05-25T15:07:47.103600148Z" level=info msg="starting containerd" revision=8c906ff108ac28da23f69cc7b74f8e7a470d1df0 version=v1.5.0
        sandbox.go:224: time="2021-05-25T15:07:47.152828214Z" level=info msg="loading plugin \"io.containerd.content.v1.content\"..." type=io.containerd.content.v1
        sandbox.go:224: time="2021-05-25T15:07:47.153541160Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.aufs\"..." type=io.containerd.snapshotter.v1
        sandbox.go:224: time="2021-05-25T15:07:47.159306931Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.devmapper\"..." type=io.containerd.snapshotter.v1
        sandbox.go:224: time="2021-05-25T15:07:47.159372835Z" level=warning msg="failed to load plugin io.containerd.snapshotter.v1.devmapper" error="devmapper not configured"
        sandbox.go:224: time="2021-05-25T15:07:47.159406437Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.native\"..." type=io.containerd.snapshotter.v1
        sandbox.go:224: time="2021-05-25T15:07:47.159489843Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.overlayfs\"..." type=io.containerd.snapshotter.v1
        sandbox.go:224: time="2021-05-25T15:07:47.159693756Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.zfs\"..." type=io.containerd.snapshotter.v1
        sandbox.go:224: time="2021-05-25T15:07:47.159935171Z" level=info msg="skip loading plugin \"io.containerd.snapshotter.v1.zfs\"..." error="path /tmp/bktest_containerd399201633/root/io.containerd.snapshotter.v1.zfs must be a zfs filesystem to be used with the zfs snapshotter: skip plugin" type=io.containerd.snapshotter.v1
        sandbox.go:224: time="2021-05-25T15:07:47.159975074Z" level=info msg="loading plugin \"io.containerd.snapshotter.v1.stargz\"..." type=io.containerd.snapshotter.v1
        sandbox.go:224: time="2021-05-25T15:07:47.160110983Z" level=info msg="loading plugin \"io.containerd.metadata.v1.bolt\"..." type=io.containerd.metadata.v1
        sandbox.go:224: time="2021-05-25T15:07:47.160537510Z" level=warning msg="could not use snapshotter devmapper in metadata plugin" error="devmapper not configured"
        sandbox.go:224: time="2021-05-25T15:07:47.160584613Z" level=info msg="metadata content store policy set" policy=shared
        sandbox.go:224: time="2021-05-25T15:07:47.163713714Z" level=info msg="loading plugin \"io.containerd.differ.v1.walking\"..." type=io.containerd.differ.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163745516Z" level=info msg="loading plugin \"io.containerd.gc.v1.scheduler\"..." type=io.containerd.gc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163781919Z" level=info msg="loading plugin \"io.containerd.service.v1.introspection-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163816221Z" level=info msg="loading plugin \"io.containerd.service.v1.containers-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163830622Z" level=info msg="loading plugin \"io.containerd.service.v1.content-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163843923Z" level=info msg="loading plugin \"io.containerd.service.v1.diff-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163857824Z" level=info msg="loading plugin \"io.containerd.service.v1.images-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163881225Z" level=info msg="loading plugin \"io.containerd.service.v1.leases-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163895626Z" level=info msg="loading plugin \"io.containerd.service.v1.namespaces-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163909627Z" level=info msg="loading plugin \"io.containerd.service.v1.snapshots-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.163926228Z" level=info msg="loading plugin \"io.containerd.runtime.v1.linux\"..." type=io.containerd.runtime.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164025034Z" level=info msg="loading plugin \"io.containerd.runtime.v2.task\"..." type=io.containerd.runtime.v2
        sandbox.go:224: time="2021-05-25T15:07:47.164111840Z" level=info msg="loading plugin \"io.containerd.monitor.v1.cgroups\"..." type=io.containerd.monitor.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164471663Z" level=info msg="loading plugin \"io.containerd.service.v1.tasks-service\"..." type=io.containerd.service.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164495965Z" level=info msg="loading plugin \"io.containerd.internal.v1.restart\"..." type=io.containerd.internal.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164541068Z" level=info msg="loading plugin \"io.containerd.grpc.v1.containers\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164555769Z" level=info msg="loading plugin \"io.containerd.grpc.v1.content\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164574570Z" level=info msg="loading plugin \"io.containerd.grpc.v1.diff\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164587671Z" level=info msg="loading plugin \"io.containerd.grpc.v1.events\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164602672Z" level=info msg="loading plugin \"io.containerd.grpc.v1.healthcheck\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164615872Z" level=info msg="loading plugin \"io.containerd.grpc.v1.images\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164627773Z" level=info msg="loading plugin \"io.containerd.grpc.v1.leases\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164639974Z" level=info msg="loading plugin \"io.containerd.grpc.v1.namespaces\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164652775Z" level=info msg="loading plugin \"io.containerd.internal.v1.opt\"..." type=io.containerd.internal.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164682077Z" level=info msg="loading plugin \"io.containerd.grpc.v1.snapshots\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164696478Z" level=info msg="loading plugin \"io.containerd.grpc.v1.tasks\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164709378Z" level=info msg="loading plugin \"io.containerd.grpc.v1.version\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.164721379Z" level=info msg="loading plugin \"io.containerd.grpc.v1.introspection\"..." type=io.containerd.grpc.v1
        sandbox.go:224: time="2021-05-25T15:07:47.165147707Z" level=info msg=serving... address=/tmp/bktest_containerd399201633/debug.sock
        sandbox.go:224: time="2021-05-25T15:07:47.165205410Z" level=info msg=serving... address=/tmp/bktest_containerd399201633/containerd.sock.ttrpc
        sandbox.go:224: time="2021-05-25T15:07:47.165247513Z" level=info msg=serving... address=/tmp/bktest_containerd399201633/containerd.sock
        sandbox.go:224: time="2021-05-25T15:07:47.165263914Z" level=debug msg="sd notification" error="<nil>" notified=false state="READY=1"
        sandbox.go:224: time="2021-05-25T15:07:47.165279915Z" level=info msg="containerd successfully booted in 0.062350s"
        sandbox.go:224: time="2021-05-25T15:07:47.271301235Z" level=debug msg="garbage collected" d=1.006264ms
        sandbox.go:224: time="2021-05-25T15:07:47.907360148Z" level=debug msg="(*service).Write started" expected="sha256:b5fc1d7b2e4ea86a06b0cf88de915a2c43a99a00b6b3c0af731e5f4c07ae8eff" ref="index-sha256:b5fc1d7b2e4ea86a06b0cf88de915a2c43a99a00b6b3c0af731e5f4c07ae8eff" total=2080
        sandbox.go:224: time="2021-05-25T15:07:47.986710653Z" level=debug msg="(*service).Write started" expected="sha256:f3cfc9d0dbf931d3db4685ec659b7ac68e2a578219da4aae65427886e649b06b" ref="manifest-sha256:f3cfc9d0dbf931d3db4685ec659b7ac68e2a578219da4aae65427886e649b06b" total=527
        sandbox.go:224: time="2021-05-25T15:07:48.038092158Z" level=debug msg="(*service).Write started" expected="sha256:d3cd072556c21c1f1940bd536675b97d7d419a2287d6bb3bd5044ea7466db788" ref="config-sha256:d3cd072556c21c1f1940bd536675b97d7d419a2287d6bb3bd5044ea7466db788" total=1457
        sandbox.go:224: time="2021-05-25T15:07:48.121287309Z" level=debug msg="stat snapshot" key="sha256:d0d0905d7be4eff6a63efe4a38647a679de1e024101f67db4fe4b5736c1e7f48"
        sandbox.go:224: time="2021-05-25T15:07:48.121943451Z" level=debug msg="(*service).Write started" expected="sha256:92f8b3f0730fef84ba9825b3af6ad90de454c4c77cde732208cf84ff7dd41208" ref="layer-sha256:92f8b3f0730fef84ba9825b3af6ad90de454c4c77cde732208cf84ff7dd41208" total=766640
        sandbox.go:224: time="2021-05-25T15:07:48.162556264Z" level=debug msg="garbage collected" d="882.657µs"
        sandbox.go:224: time="2021-05-25T15:07:48.195617390Z" level=debug msg="prepare snapshot" key="extract-mcldenwtueew6cztyr43g47jr sha256:d0d0905d7be4eff6a63efe4a38647a679de1e024101f67db4fe4b5736c1e7f48" parent=
        sandbox.go:224: time="2021-05-25T15:07:48.201887893Z" level=debug msg="event published" ns=buildkit-test topic=/snapshot/prepare type=containerd.events.SnapshotPrepare
        sandbox.go:224: time="2021-05-25T15:07:48.202679744Z" level=debug msg="get snapshot mounts" key="extract-mcldenwtueew6cztyr43g47jr sha256:d0d0905d7be4eff6a63efe4a38647a679de1e024101f67db4fe4b5736c1e7f48"
        sandbox.go:224: time="2021-05-25T15:07:48.203431593Z" level=debug msg="unpigz not found, falling back to go gzip" error="exec: \"unpigz\": executable file not found in $PATH"
        sandbox.go:224: time="2021-05-25T15:07:48.312964538Z" level=debug msg="diff applied" d=109.659554ms digest="sha256:92f8b3f0730fef84ba9825b3af6ad90de454c4c77cde732208cf84ff7dd41208" media=application/vnd.docker.image.rootfs.diff.tar.gzip size=766640
        sandbox.go:224: time="2021-05-25T15:07:48.313738488Z" level=debug msg="commit snapshot" key="extract-mcldenwtueew6cztyr43g47jr sha256:d0d0905d7be4eff6a63efe4a38647a679de1e024101f67db4fe4b5736c1e7f48" name="sha256:d0d0905d7be4eff6a63efe4a38647a679de1e024101f67db4fe4b5736c1e7f48"
        sandbox.go:224: time="2021-05-25T15:07:48.321589193Z" level=debug msg="event published" ns=buildkit-test topic=/snapshot/commit type=containerd.events.SnapshotCommit
        sandbox.go:224: time="2021-05-25T15:07:48.330393659Z" level=debug msg="prepare snapshot" key=cg46l8ymxhobcsp798nqcrbuw parent="sha256:d0d0905d7be4eff6a63efe4a38647a679de1e024101f67db4fe4b5736c1e7f48"
        sandbox.go:224: time="2021-05-25T15:07:48.334153901Z" level=debug msg="event published" ns=buildkit-test topic=/snapshot/prepare type=containerd.events.SnapshotPrepare
        sandbox.go:224: time="2021-05-25T15:07:48.336853075Z" level=debug msg="get snapshot mounts" key=cg46l8ymxhobcsp798nqcrbuw
        sandbox.go:224: time="2021-05-25T15:07:48.343887827Z" level=debug msg="event published" ns=buildkit-test topic=/containers/create type=containerd.events.ContainerCreate
        sandbox.go:224: time="2021-05-25T15:07:48.348830945Z" level=debug msg="garbage collected" d="815.252µs"
        sandbox.go:224: time="2021-05-25T15:07:48.365171897Z" level=info msg="starting signal loop" namespace=buildkit-test path=/tmp/bktest_containerd399201633/state/io.containerd.runtime.v2.task/buildkit-test/wd4kni1y3hihiyz7gawutpz9l pid=9837
        sandbox.go:224: time="2021-05-25T15:07:48.434257440Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/create type=containerd.events.TaskCreate
        sandbox.go:224: time="2021-05-25T15:07:48.444915526Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/start type=containerd.events.TaskStart
        sandbox.go:224: time="2021-05-25T15:07:48.454085716Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/exit type=containerd.events.TaskExit
        sandbox.go:224: time="2021-05-25T15:07:48.492705200Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/delete type=containerd.events.TaskDelete
        sandbox.go:224: time="2021-05-25T15:07:48.494404009Z" level=info msg="shim disconnected" id=wd4kni1y3hihiyz7gawutpz9l
        sandbox.go:224: time="2021-05-25T15:07:48.494436411Z" level=warning msg="cleaning up after shim disconnected" id=wd4kni1y3hihiyz7gawutpz9l namespace=buildkit-test
        sandbox.go:224: time="2021-05-25T15:07:48.494445112Z" level=info msg="cleaning up dead shim"
        sandbox.go:224: time="2021-05-25T15:07:48.538200126Z" level=warning msg="cleanup warnings time=\"2021-05-25T15:07:48Z\" level=info msg=\"starting signal loop\" namespace=buildkit-test pid=9962\n"
        sandbox.go:224: time="2021-05-25T15:07:48.540705887Z" level=debug msg="event published" ns=buildkit-test topic=/containers/delete type=containerd.events.ContainerDelete
        sandbox.go:224: time="2021-05-25T15:07:48.544014300Z" level=debug msg="get snapshot mounts" key=cg46l8ymxhobcsp798nqcrbuw
        sandbox.go:224: time="2021-05-25T15:07:48.547020794Z" level=debug msg="event published" ns=buildkit-test topic=/containers/create type=containerd.events.ContainerCreate
        sandbox.go:224: time="2021-05-25T15:07:48.576220272Z" level=debug msg="garbage collected" d="676.944µs"
        sandbox.go:224: time="2021-05-25T15:07:48.583951669Z" level=info msg="starting signal loop" namespace=buildkit-test path=/tmp/bktest_containerd399201633/state/io.containerd.runtime.v2.task/buildkit-test/wd4kni1y3hihiyz7gawutpz9l pid=10026
        sandbox.go:224: time="2021-05-25T15:07:48.643406094Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/create type=containerd.events.TaskCreate
        sandbox.go:224: time="2021-05-25T15:07:48.664192831Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/start type=containerd.events.TaskStart
        sandbox.go:224: time="2021-05-25T15:07:48.670988368Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/exec-added type=containerd.events.TaskExecAdded
        sandbox.go:224: time="2021-05-25T15:07:48.730630604Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/exec-started type=containerd.events.TaskExecStarted
        sandbox.go:224: time="2021-05-25T15:07:48.731091234Z" level=debug msg="event forwarded" ns=buildkit-test topic=/tasks/exit type=containerd.events.TaskExit
        sandbox.go:224: > sending sigterm 2021-05-25 15:07:48.733705002 +0000 UTC m=+1.842541720
        sandbox.go:224: time="2021-05-25T15:07:48.733768106Z" level=debug msg="received signal" signal=terminated
        sandbox.go:224: time="2021-05-25T15:07:48.733815209Z" level=debug msg="sd notification" error="<nil>" notified=false state="STOPPING=1"
        sandbox.go:224: > stopped 2021-05-25 15:07:48.736045052 +0000 UTC m=+1.844881870 exit status 0 0
        sandbox.go:221: stdout: /usr/bin/buildkitd
        sandbox.go:221: stderr: /usr/bin/buildkitd
        sandbox.go:224: > startCmd 2021-05-25 15:07:47.202435505 +0000 UTC m=+0.311272223 [buildkitd --oci-worker=false --containerd-worker-gc=false --containerd-worker=true --containerd-worker-addr /tmp/bktest_containerd399201633/containerd.sock --containerd-worker-labels=org.mobyproject.buildkit.worker.sandbox=true --containerd-worker-snapshotter=stargz --config=/tmp/bktest_config300765578/buildkitd.toml --root /tmp/bktest_buildkitd143239262 --addr unix:///tmp/bktest_buildkitd143239262/buildkitd.sock --debug]
        sandbox.go:224: time="2021-05-25T15:07:47Z" level=debug msg="remote introspection plugin filters" filters="[type==io.containerd.runtime.v1 type==io.containerd.runtime.v2]"
        sandbox.go:224: time="2021-05-25T15:07:47Z" level=info msg="found worker \"to64zr7yc4scht6bdyh8f37m4\", labels=map[org.mobyproject.buildkit.worker.containerd.namespace:buildkit org.mobyproject.buildkit.worker.containerd.uuid:e23955f3-2133-4699-a22a-74b4ff84eda6 org.mobyproject.buildkit.worker.executor:containerd org.mobyproject.buildkit.worker.hostname:aafe8f8c0ad5 org.mobyproject.buildkit.worker.sandbox:true org.mobyproject.buildkit.worker.snapshotter:stargz], platforms=[linux/amd64 linux/arm64 linux/riscv64 linux/ppc64le linux/s390x linux/386 linux/mips64le linux/mips64 linux/arm/v7 linux/arm/v6]"
        sandbox.go:224: time="2021-05-25T15:07:47Z" level=info msg="found 1 workers, default=\"to64zr7yc4scht6bdyh8f37m4\""
        sandbox.go:224: time="2021-05-25T15:07:47Z" level=warning msg="currently, only the default worker can be used."
        sandbox.go:224: time="2021-05-25T15:07:47Z" level=info msg="running server on /tmp/bktest_buildkitd143239262/buildkitd.sock"
        sandbox.go:224: > sending sigterm 2021-05-25 15:07:48.73228281 +0000 UTC m=+1.841119528
        sandbox.go:224: time="2021-05-25T15:07:48Z" level=info msg="stopping server"
        sandbox.go:224: buildkitd: context canceled
        sandbox.go:224: > stopped 2021-05-25 15:07:48.733384781 +0000 UTC m=+1.842221599 exit status 1 1
    --- PASS: TestContainerdWorkerIntegration/TestContainerdWorkerExec/worker=containerd (1.85s)
        common.go:78: Stdout: hello
        common.go:79: Stderr: 
        common.go:114: Stdout: PID   COMMAND
                1 sleep
                7 ps
        common.go:115: Stderr: 
        common.go:147: Stdout: hello
        common.go:148: Stderr: 
 
```

@coryb

--- a/executor/containerdexecutor/executor.go
+++ b/executor/containerdexecutor/executor.go
@@ -330,6 +330,12 @@ func (w *containerdExecutor) runProcess(ctx context.Context, p containerd.Proces
 		return err
 	}
 
+	io := p.IO()
+	defer func() {
+		io.Wait()
+		io.Close()
+	}()
+
 	err = p.Start(ctx)
 	if err != nil {
 		return err
@@ -373,6 +379,7 @@ func (w *containerdExecutor) runProcess(ctx context.Context, p containerd.Proces
 			killCtx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
 			killCtxDone = killCtx.Done()
 			p.Kill(killCtx, syscall.SIGKILL)
+			io.Cancel()
 		case status := <-statusCh:
 			if cancel != nil {
 				cancel()
@@ -397,6 +404,7 @@ func (w *containerdExecutor) runProcess(ctx context.Context, p containerd.Proces
 			if cancel != nil {
 				cancel()
 			}
+			io.Cancel()
 			return errors.Errorf("failed to kill process on cancel")
 		}
 	}

